disaster recovery guide for PVCs (#6452)
rook/rook: Documentation/ceph-disaster-recovery.md

1. Bring the Rook Ceph operator back online by running `kubectl -n rook-ceph edit deploy/rook-ceph-operator` and set `replicas` to `1`.
1. Watch the operator logs with `kubectl -n rook-ceph logs -f rook-ceph-operator-xxxxxxx`, and wait until the orchestration has settled.
1. **STATE**: Now the new cluster should be up and running with authentication enabled. `ceph -s` output should not change much comparing to previous steps.

> 

Reviewer: #4838 was written with the intention to restore a Rook cluster if everything was lost except the mon and osd PVCs, but there were a lot of open questions about how to do that. Perhaps it's better to abandon that PR and go with the assumption that you must back up the required resources if you expect to survive recovery. @tareqhs would you mind taking a look at that PR and see if there is anything useful to add here?

Author: is your intention to merge this into #4838 ?

Reviewer: I went ahead and closed #4838 in favor of this one where we assume the metadata is backed up.

Author: thanks

## Adopting into a new Kubernetes cluster with PVCs

> It is possible to migrate/restore an rook/ceph cluster from an exising Kubernetes cluster to a new one without restoriting to SSH access or ceph tooling. This allows doing the migration using standard kubernetes resources only. This guide assumes you have a CephCluster that uses PVCs to persist mon and osd data.

Reviewer 2: typo: restoriting, perhaps you meant "resorting"?

> 1. Stop rook in the cluster cluster by scaling the operator deployment `rook-ceph-operator` down to zero and deleting the other deployments: `rook-ceph-mgr-a`, `rook-ceph-mon-a`, ..., `rook-ceph-osd-0`, ...

Reviewer 3: ```suggestion
1. Stop rook in the old cluster by scaling the operator deployment `rook-ceph-operator` down to zero and deleting the other deployments: `rook-ceph-mgr-a`, `rook-ceph-mon-a`, ..., `rook-ceph-osd-0`, ...
```

Author: I actually meant the new cluster. restructured the description to make it clearer.

> 1. Restore the rook PVCs to the new cluster. A common way to do this is to generate regular snapshots of the volumes and re-create PVCs from these volumes in the new cluster.

Reviewer: Should we mention Velero as an example to back them up?

Author: Added. I tried to avoid this and leave the user to pick any tool without preferring one over another.

> 1. Copy the keyring and fsid secretsfrom the old cluster: `rook-ceph-mgr-a-keyring`, `rook-ceph-mon`, `rook-ceph-mons-keyring`, `rook-ceph-osd-0-keyring`, ...

Reviewer 3: ```suggestion
1. Copy the keyring and fsid secrets from the old cluster: `rook-ceph-mgr-a-keyring`, `rook-ceph-mon`, `rook-ceph-mons-keyring`, `rook-ceph-osd-0-keyring`, ...
```

> 1. Delete mon services and copy them from the old cluster: `rook-ceph-mon-a`, `rook-ceph-mon-b`, ... Note the simply re-applying won't work because the goal here is to restore the `clusterIP` in each service and this field is immutable in `Service` resources.

Reviewer 3: ```suggestion
1. Delete mon services and copy them from the old cluster: `rook-ceph-mon-a`, `rook-ceph-mon-b`, ... Note that simply re-applying won't work because the goal here is to restore the `clusterIP` in each service and this field is immutable in `Service` resources.
```

1. Copy the endpoints configmap from the old cluster: `rook-ceph-mon-endpoints`

> 1 Scale the operator back up and wait until the reconciliation is over.

Reviewer 2: need "1."

> 

Reviewer: How about showing kubectl or other commands that will help users get this working? Right now the document is high level and may be ambiguous for how to backup and restore the resources.

Reviewer: I see your point that the exact commands will depend on the tools. This is an advanced scenario, so it seems acceptable to leave the details to the reader.

Author: Actually I did add several kubectl commands as examples..

1. Bring the Rook Ceph operator back online by running `kubectl -n rook-ceph edit deploy/rook-ceph-operator` and set `replicas` to `1`.
1. Watch the operator logs with `kubectl -n rook-ceph logs -f rook-ceph-operator-xxxxxxx`, and wait until the orchestration has settled.
1. **STATE**: Now the new cluster should be up and running with authentication enabled. `ceph -s` output should not change much comparing to previous steps.

> ## Adopting into a new Kubernetes cluster with PVCs

Reviewer: What about this title, and also moving it higher in the document before the section "Adopt an existing Rook Ceph cluster into a new Kubernetes cluster"? I think this will be a more useful/common scenario than the other one. Maybe we need a simple list of links at the top to the three scenarios as well since it's getting longer, similar to the [common issues](https://rook.github.io/docs/rook/v1.4/ceph-common-issues.html) page.
```suggestion
## Backing up and restoring a cluster based on PVCs into a new Kubernetes cluster
```

1. Bring the Rook Ceph operator back online by running `kubectl -n rook-ceph edit deploy/rook-ceph-operator` and set `replicas` to `1`.
1. Watch the operator logs with `kubectl -n rook-ceph logs -f rook-ceph-operator-xxxxxxx`, and wait until the orchestration has settled.
1. **STATE**: Now the new cluster should be up and running with authentication enabled. `ceph -s` output should not change much comparing to previous steps.

## Adopting into a new Kubernetes cluster with PVCs

It is possible to migrate/restore an rook/ceph cluster from an exising Kubernetes cluster to a new one without resorting to SSH access or ceph tooling. This allows doing the migration using standard kubernetes resources only. This guide assumes the following
1. You have a CephCluster that uses PVCs to persist mon and osd data. Let's call it the "old cluster"

> 1. You can restore the PVCs as-is in the new cluster. Usually this is done by taking regular snapshots of the PVC volumes and using a tool that can re-create PVCs from these snapshots in the underlying cloud provider. Velero one such tool. (https://github.com/vmware-tanzu/velero)

Reviewer: ```suggestion
1. You can restore the PVCs as-is in the new cluster. Usually this is done by taking regular snapshots of the PVC volumes and using a tool that can re-create PVCs from these snapshots in the underlying cloud provider. Velero is one such tool. (https://github.com/vmware-tanzu/velero)
```

1. You have regular backups of the secrets and configmaps in the rook-ceph namespace. Velero provides this functionality too.

Do the following in the new cluster:
1. Stop the rook operator by scaling the deployment `rook-ceph-operator` down to zero: `kubectl -n rook-ceph scale deployment rook-ceph-operator --replicas 0`

> and deleting the other deployments. An example command to do this is `k -n rook-ceph delete deployment -l operator!=rook`

Reviewer: Why not delete the operator deployment as well? Scaling it down to zero isn't much different than deleting the deployment. Or else why not scale down all the deployments to 0 instead of deleting them?

Author: User convenience. All the other deployments are created and reconciled by operator so once the operator is scaled up they will be restored automatically. The operator itself is installed as part of rook and deleting it will require re-installing rook-ceph (helm, argo, ...).

Reviewer: Agreed, when installed with the helm chart you can't just remove the deployment and re-create it easily.
